pkg/result/processors/sort_results.go

```diff
@@ -0,0 +1,179 @@
+// by sorting results.Issues using processor step, and chain based
+// rules that can compare different properties of the Issues struct.
+
+var _ Processor = (*SortResults)(nil)
+
+type SortResults struct {
+	cmp Comparator
+	cfg *config.Config
+}
+
+func NewSortResults(cfg *config.Config) *SortResults {
+	// For sorting we are comparing (in next order): file names, line numbers,
+	// position, range numbers (if position not available), and finally - giving up.
+	return &SortResults{
+		cmp: ByName{
+			next: ByLine{
+				next: ByColumn{},
+			},
+		},
+		cfg: cfg,
+	}
+}
+
+// Process is performing sorting of the result issues.
+func (sr SortResults) Process(issues []result.Issue) ([]result.Issue, error) {
+	if !sr.cfg.Output.SortResults {
+		return issues, nil
+	}
+
+	sort.Slice(issues, func(i, j int) bool {
+		return sr.cmp.Compare(&issues[i], &issues[j]) == Less
+	})
+
+	return issues, nil
+}
+
+func (sr SortResults) Name() string { return "sort_results" }
+func (sr SortResults) Finish()      {}
+
+type CompareResult int
+
+const (
+	Less CompareResult = iota - 1
+	Equal
+	Greater
+	None
+)
+
+func (c CompareResult) isNeutral() bool {
+	// return true if compare result is incomparable or equal.
+	return c == None || c == Equal
+}
+
+func (c CompareResult) String() string {
+	switch c {
+	case Less:
+		return "Less"
+	case Equal:
+		return "Equal"
+	case Greater:
+		return "Greater"
+	}
+
+	return "None"
+}
+
+// Comparator describe how to implement compare for two "issues" lexicographically
+type Comparator interface {
+	Compare(a, b *result.Issue) CompareResult
+	Next() Comparator
+}
+
+var (
+	_ Comparator = (*ByName)(nil)
+	_ Comparator = (*ByLine)(nil)
+	_ Comparator = (*ByColumn)(nil)
+	_ Comparator = (*ByRange)(nil)
+)
+
+type ByName struct{ next Comparator }
+
+func (cmp ByName) Next() Comparator { return cmp.next }
+func (cmp ByName) Compare(a, b *result.Issue) CompareResult {
+	var res CompareResult
+
+	if res = CompareResult(strings.Compare(a.FilePath(), b.FilePath())); !res.isNeutral() {
+		return res
+	}
+
+	if next := cmp.Next(); next != nil {
+		return next.Compare(a, b)
+	}
+
+	return res
+}
+
+type ByLine struct{ next Comparator }
+
+func (cmp ByLine) Next() Comparator { return cmp.next }
+func (cmp ByLine) Compare(a, b *result.Issue) CompareResult {
+	var res CompareResult
+
+	if res = numericCompare(a.Line(), b.Line()); !res.isNeutral() {
+		return res
+	} else if next := cmp.Next(); res.isNeutral() && next != nil {
+		return next.Compare(a, b)
+	}
+	return res
+}
+
+type ByColumn struct{ next Comparator }
+
+func (cmp ByColumn) Next() Comparator { return cmp.next }
+func (cmp ByColumn) Compare(a, b *result.Issue) CompareResult {
+	var res CompareResult
+
+	if res = numericCompare(a.Column(), b.Column()); !res.isNeutral() {
+		return res
+	} else if next := cmp.Next(); res.isNeutral() && next != nil {
+		return next.Compare(a, b)
+	}
+
+	return res
+}
+
+type ByRange struct{ next Comparator }
```

Comment:
Just want to confirm if this one is in used.

Reply:
It is. But I have trouble to run the demo (not sure is that a `golangci-lint` bug, or just my configuration).

The base idea, that linters can generate different output, in purchase fallback to `ByRange` can happen if there a duplication code in the few copy/paste files in different packages, so output can looks like this.

```
examples/dupl1.go:1: 1-15 lines are duplicate of `examples/dupl2.go:1-15` (dupl)
examples/dupl1.go:1: 1-14 lines are duplicate of `examples/dupl3.go:1-14` (dupl)
examples/dupl2.go:1: 1-15 lines are duplicate of `examples/dupl3.go:1-15` (dupl)
examples/dupl3.go:1: 1-15 lines are duplicate of `examples/dupl1.go:1-15` (dupl)
...
```

Correct me if I am wrong and such output is impossible.

Reply:
Oh, so you tried to add _ByRange_ but it's not working as expected right? The current rules are initialized as below only

```
	return &SortResults{
		cmp: ByName{
			next: ByLine{
				next: ByColumn{},
			},
		},
		cfg: cfg,
	}
```

I am ok to remove this ByRange comparator, as sorting till Line number is good enough for me though. It's best effort anyway right now, one benefit is that we can simplify numericCompare or remove this method.

Reply:
I did delete ByRange till the moment I can confirm the results of `dupl` linter are parsed correctly (or not) - that's the only case its can be useful afaik.